src/copy.cpp

```diff
@@ -0,0 +1,98 @@
+#include <filesystem>
+#include <fstream>
+#include <vector>
+#include <iostream>
+#include <thread>
+#include <mutex>
+#include <copy.h>
+#include <exceptions/io_exception.h>
+
+std::exception_ptr Copy::exceptionPtr = nullptr;
+
+Copy::Copy(CopyArgumentParser args) : _args(std::move(args)), _done_reading(false) {
+    std::cout << "Try to copy " << _args.sourcePath() << " to " << _args.destinationPath() << std::endl;
+}
+
+void Copy::runOneThread(const int buffer_size) const {
+
+    std::ifstream fin(_args.sourcePath(), std::ifstream::binary);
+    if (!fin.is_open())
+        throw CopyFileReadException(_args.sourcePath());
+
+    std::ofstream fout(_args.destinationPath(), std::ifstream::binary);
+    if (!fout.is_open())
+        throw CopyFileWriteException(_args.destinationPath());
+
+    std::unique_ptr<char[]> buffer(new char[buffer_size]);
+
+    while (fin.good()) {
+        fin.read(buffer.get(), buffer_size);
+        std::streamsize s = fin.gcount();
+
+        if (fout.good())
+            fout.write(buffer.get(), s);
+        else
+            throw CopyFileWriteException(_args.sourcePath());
+    }
+}
+
+void Copy::parallelCopy(const int buffer_size) {
+    std::thread reader(&Copy::reader, this, buffer_size);
+    std::thread writer(&Copy::writer, this);
+
+    reader.join();
+    writer.join();
+}
+
+void Copy::writer() {
+    try {
+        std::ofstream fout(_args.destinationPath());
+        if (!fout.is_open())
+            throw CopyFileWriteException(_args.destinationPath());
+
+        while (!_done_reading || !_buffer_queue.empty()) {
+            while (!_buffer_queue.empty()) {
+                std::pair<std::unique_ptr<char []>, int> buffer_obj;
```

Comment:
you should always initialize values. Here, the buffer_obj.second is initialized with a random value. 
Prefer constant objects always, instead of the creation of default objects + later assigning them to some valid one.
There are many ways to make it better: 
- use std::optional
- call a function that returns std::pair<std::unique_ptr<char []>, int>  and initialize the buffer_obj by calling the function.